src/basic/hostname-util.c

```diff
@@ -221,6 +221,39 @@ int sethostname_idempotent(const char *s) {
         return 1;
 }
 
+int shorten_overlong(const char *s, char **ret) {
+        char *h, *p;
+
+        /* Shorten an overlong name to HOST_NAME_MAX or to the first dot,
+         * whatever comes earlier. */
+
+        assert(s);
+
+        h = strdup(s);
+        if (!h)
+                return -ENOMEM;
+
+        if (strlen(h) <= HOST_NAME_MAX) {
+                *ret = h;
+                return 0;
+        }
+
+        for (p = h; *p; p++) {
+                if (*p == '.') {
+                        *p = 0;
+                        break;
+                }
+        }
+
+        strshorten(h, HOST_NAME_MAX);
+
+        if (strlen(h) == 0)
```

Comment:
strlen() is O(n). If all you want to check is whether the string is empty, is "isempty()" instead...

That said, i think it would be better to instead just call hostname_is_valid(…, false) here